libyul/AsmParser.cpp

```diff
@@ -78,14 +93,67 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
 	return nullptr;
 }
 
+langutil::Token Parser::advance()
+{
+	auto const token = ParserBase::advance();
+	if (m_useSourceLocationFrom == UseSourceLocationFrom::Comments)
+		fetchSourceLocationFromComment();
+	return token;
+}
+
+void Parser::fetchSourceLocationFromComment()
+{
+	solAssert(m_charStreamMap.has_value(), "");
+
+	if (m_scanner->currentCommentLiteral().empty())
+		return;
+
+	static std::regex const lineRE = std::regex(
+		R"~~~((^|\s+)@src\s+(-1|\d+):(-1|\d+):(-1|\d+)(\s+|$))~~~",
+		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	);
+
+	string text = m_scanner->currentCommentLiteral();
+	while (!text.empty())
+	{
+		std::smatch matchResult;
+		if (!std::regex_search(text, matchResult, lineRE))
+			return;
+		solAssert(matchResult.size() == 6, "");
+
+		auto const sourceIndex = toInt(matchResult[2].str());
+		auto const start = toInt(matchResult[3].str());
+		auto const end = toInt(matchResult[4].str());
+
+		auto const commentLocation = m_scanner->currentCommentLocation();
+		m_locationOverride = SourceLocation{};
+		if (!sourceIndex || !start || !end)
+			m_errorReporter.syntaxError(6367_error, commentLocation, "Invalid value in source location mapping. Could not parse location specification.");
+		else if (!((start < 0 && end < 0) || (start >= 0 && *start <= *end)))
+			m_errorReporter.syntaxError(5798_error, commentLocation, "Invalid value in source location mapping. Start offset larger than end offset.");
+		else if (!(sourceIndex >= 0 && m_charStreamMap->count(static_cast<unsigned>(*sourceIndex))))
+			m_errorReporter.syntaxError(2674_error, commentLocation, "Invalid source mapping. Source index not defined via @use-src.");
+		else if (sourceIndex >= 0)
```

Comment:
I'm wondering if we should still allow `-1:0:10` to lead to a "real" source location (instead of the default-constructed one)

Reply:
hm, I think that sounds like an interesting thing. I can do that. It would again require some more test cases, I think.

Reply:
Then please do.